File: 5.电话漏接分析/电话漏接分析后台/missed_call_backend/tui_app.py
"""电话漏接分析的全屏 TUI 框架。

参考“客服超时督办”的 TUI 交互：顶部标题、状态栏、菜单栏、内容区、
底部快捷键提示，支持上下左右键切换和选择。本模块只负责终端框架与页面调度，
具体业务页面由页面对象提供 render/handle_key。
"""
from __future__ import annotations

import os
import shutil
import sys
import time
from typing import Any, Callable, Sequence

from . import cli_display
from .cli_display import (
    colorize,
    fit_text,
)

# 游戏化提示音总开关：False 时所有操作不发声（仅保留视觉反馈）。
SOUND_ENABLED = True


def _fit(text: object, width: int, truncate: bool = True) -> str:
    """把一行文本按终端宽度补齐。

    truncate=False 时只补短行、不截长行，用于表格等内容完整展示。
    """
    return fit_text(text, width, truncate)


class Page:
    """TUI 页面基类。"""

    key = ""
    title = ""
    # 页面存在后台持续变化（如下载进度）时置为 True，空闲时也保持定时刷新；
    # 其余页面空闲时完全不重绘，避免打断鼠标选择复制。
    periodic_refresh = False

    def __init__(self) -> None:
        self.state: dict[str, Any] = {}

    def on_enter(self, app: "TuiApp") -> None:
        """进入页面时回调，可用来重置滚动或重新加载数据。"""

    def render(self, app: "TuiApp") -> list[str]:
        """返回内容区文本行，不要包含终端控制序列以外的格式。"""
        return []

    def handle_key(self, key: str, app: "TuiApp") -> bool:
        """处理按键，返回 True 表示已消费。"""
        return False

    def footer(self, app: "TuiApp") -> str:
        """返回底部快捷键提示。"""
        return ""


class TuiApp:
    """全屏 TUI 应用：负责原始按键、备用屏幕、帧渲染和页面切换。"""

    def __init__(
        self,
        title: str,
        pages: Sequence[Page],
        application: Any,
        status_provider: Callable[["TuiApp"], list[str]] | None = None,
        output: Any = None,
    ) -> None:
        self.title = title
        self.pages = list(pages)
        self.application = application
        self.status_provider = status_provider or (lambda _app: [])
        self.output = output or sys.stdout
        self.current_page_index = 0
        self.running = False
        self._old_termios: Any = None
        self._last_output = ""
        self._render_requested = False
        self._toast = ""
        self._toast_at = 0.0

    @property
    def page(self) -> Page | None:
        if 0 <= self.current_page_index < len(self.pages):
            return self.pages[self.current_page_index]
        return None

    @property
    def columns(self) -> int:
        # 优先读输出对象自带的宽度（测试桩等），否则查询真实终端宽度。
        # 注意不要用 self.output.columns 直接取值：sys.stdout 没有该属性，
        # 直接访问会抛 AttributeError 导致宽度永远退化为 80，选中行高亮铺不满整行。
        width = getattr(self.output, "columns", 0) or 0
        if width <= 0:
            try:
                width = shutil.get_terminal_size(fallback=(80, 24)).columns
            except (AttributeError, OSError, ValueError):
                width = 80
        return max(48, width)

    @property
    def rows(self) -> int:
        height = getattr(self.output, "rows", 0) or 0
        if height <= 0:
            try:
                height = shutil.get_terminal_size(fallback=(80, 24)).lines
            except (AttributeError, OSError, ValueError):
                height = 24
        return max(14, height)

    @property
    def content_height(self) -> int:
        # 标题 1 + 状态 2 + 菜单 1 + 分隔 1 + 联系方式 1 + 版权 1 + 页脚 1 + 底边 1 = 9 行
        return max(4, self.rows - 9)

    def start(self) -> None:
        if self.running:
            return
        self.running = True
        cli_display.COLOR_OUTPUT_ENABLED = True
        self.output.write("\x1b[?1049h\x1b[2J\x1b[?25l")
        if os.name != "nt" and hasattr(sys.stdin, "fileno") and sys.stdin.isatty():
            import termios
            import tty

            self._old_termios = termios.tcgetattr(sys.stdin.fileno())
            tty.setraw(sys.stdin.fileno())
        if self.page and hasattr(self.page, "on_enter"):
            self.page.on_enter(self)
        self.render()

    def stop(self) -> None:
        if not self.running:
            return
        self.running = False
        if self._old_termios is not None:
            import termios

            termios.tcsetattr(sys.stdin.fileno(), termios.TCSADRAIN, self._old_termios)
            self._old_termios = None
        self.output.write("\x1b[0m\x1b[?25h\x1b[?1049l")
        try:
            self.output.flush()
        except (AttributeError, OSError):
            pass

    def request_render(self) -> None:
        """请求主循环在下一次空闲轮询时刷新，供后台加载线程完成后使用。"""
        self._render_requested = True

    def feedback(self, message: str, beep: int | None = 880) -> None:
        """游戏化操作反馈：显示一条短暂提示，并可选播放提示音。

        beep 为提示音频率(Hz)，None 表示只显示提示不发声。
        """
        self.toast(message)
        if beep:
            self._beep(int(beep), 50)

    def toast(self, message: str) -> None:
        """记录一条操作提示，build_frame 会在状态区短暂显示。"""
        self._toast = message
        self._toast_at = time.monotonic()

    def _beep(self, frequency: int, duration: int) -> None:
        """Windows 提示音；非真实终端或播放失败时静默，不打断测试环境。"""
        if not SOUND_ENABLED:
            return
        isatty = getattr(self.output, "isatty", None)
        if isatty is not None and not isatty():
            return
        try:
            if os.name == "nt":
                import winsound

                winsound.Beep(frequency, duration)
        except Exception:
            pass

    def switch_page(self, index: int) -> None:
        target = int(index)
        if 0 <= target < len(self.pages):
            if target != self.current_page_index:
                self.current_page_index = target
                self.feedback(f"进入：{self.page.title}", beep=660)
            if self.page and hasattr(self.page, "on_enter"):
                self.page.on_enter(self)

    def run(self) -> None:
        self.start()
        try:
            while self.running:
                key = self.read_key(timeout=0.2)
                if key is None:
                    # 空闲时不要反复清屏重绘：会把屏幕内容不断刷新，导致
                    # 鼠标无法选中文本、复制失效。后台任务完成时可主动请求一次刷新；
                    # 其余页面只有显式 periodic_refresh 才继续定时刷新。
                    if self._render_requested:
                        self._render_requested = False
                        self.render()
                    elif self.page and self.page.periodic_refresh:
                        self.render()
                    continue
                self.dispatch_key(key)
                self.render()
        except KeyboardInterrupt:
            pass
        finally:
            self.stop()

    def read_key(self, timeout: float | None = None) -> str | None:
        """读取一个按键，Windows 使用 msvcrt，Linux/macOS 使用 termios 原始模式。

        timeout 大于 0 时最多等待这么多秒；超时返回 None，方便下载页定时刷新。
        """
        if os.name == "nt":
            import msvcrt
            import time as _time

            start = _time.monotonic()
            while True:
                if msvcrt.kbhit():
                    ch = msvcrt.getwch()
                    if ch in ("\x00", "\xe0"):
                        ch2 = msvcrt.getwch()
                        return {
                            "H": "up",
                            "P": "down",
                            "K": "left",
                            "M": "right",
                            "G": "home",
                            "O": "end",
                            "I": "pgup",
                            "Q": "pgdn",
                            "S": "delete",
                        }.get(ch2, ch2)
                    if ch == "\r":
                        return "enter"
                    if ch == "\x03":
                        return "ctrl-c"
                    if ch == "\x1b":
                        return "esc"
                    if ch in ("\x08", "\x7f"):
                        return "backspace"
                    if ch == "\t":
                        return "tab"
                    return ch
                if timeout is not None and _time.monotonic() - start >= timeout:
                    return None
                _time.sleep(0.03)

        # Unix 原始模式下的简易读取：支持方向键和常见控制键。
        import select
        import termios

        fd = sys.stdin.fileno()
        if not sys.stdin.isatty():
            return None
        ready, _, _ = select.select([fd], [], [], timeout if timeout is not None else 0.2)
        if not ready:
            return None
        data = os.read(fd, 1)
        if not data:
            return None
        char = data.decode("utf-8", errors="ignore")
        if char == "\x1b":
            # 尝试读取完整的 CSI 序列
            try:
                ready, _, _ = select.select([fd], [], [], 0.05)
                if not ready:
                    return "esc"
                seq = os.read(fd, 2).decode("utf-8", errors="ignore")
                if seq == "[A":
                    return "up"
                if seq == "[B":
                    return "down"
                if seq == "[C":
                    return "right"
                if seq == "[D":
                    return "left"
                if seq == "[H":
                    return "home"
                if seq == "[F":
                    return "end"
                if seq == "[5~":
                    return "pgup"
                if seq == "[6~":
                    return "pgdn"
                return "unknown"
            except OSError:
                return "esc"
        if char == "\r":
            return "enter"
        if char == "\x03":
            return "ctrl-c"
        if char in ("\x7f", "\x08"):
            return "backspace"
        if char == "\t":
            return "tab"
        return char

    def dispatch_key(self, key: str) -> None:
        if not key or key == "unknown":
            return

        if key == "ctrl-c":
            self.running = False
            return

        if self.page and self.page.handle_key(key, self):
            return

        if key in ("left", "right"):
            direction = 1 if key == "right" else -1
            self.switch_page((self.current_page_index + direction + len(self.pages)) % len(self.pages))
            return

        if key.isdigit():
            number = int(key)
            if 1 <= number <= len(self.pages):
                self.switch_page(number - 1)
                return

        if key == "q":
            self.switch_page(0)

    def build_menu_bar(self, columns: int) -> str:
        segments = []
        for index, page in enumerate(self.pages):
            label = f" {index + 1}{page.title} "
            if index == self.current_page_index:
                segments.append(colorize(label, "reverse"))
            else:
                segments.append(colorize(label, "brightBlue"))
        return _fit("".join(segments), columns)

    def build_frame(self) -> list[str]:
        columns = self.columns
        rows = self.rows
        content_height = self.content_height
        page = self.page
        page_title = f"{page.key}.{page.title}" if page else ""

        lines: list[str] = []
        title_bar = f" {self.title}  {('│ ' + page_title) if page_title else ''}"
        clock = time.strftime("%Y-%m-%d %H:%M:%S")
        lines.append(
            colorize(_fit(title_bar, columns - 21), "brightCyan")
            + colorize(_fit(f" {clock}", 21), "gray")
        )

        status_lines = self.status_provider(self) or []
        toast = ""
        if self._toast and time.monotonic() - self._toast_at < 1.5:
            toast = colorize("▶ " + self._toast, "brightYellow")
        for index in range(2):
            if index == 1 and toast:
                lines.append(_fit(toast, columns))
            else:
                lines.append(_fit(status_lines[index] if index < len(status_lines) else "", columns))

        lines.append(self.build_menu_bar(columns))
        lines.append(colorize("─" * columns, "gray"))

        content_lines = page.render(self) if page else []
        for content_line in content_lines:
            lines.append(_fit(content_line, columns, truncate=False))

        lines.append(colorize(_fit("作者：黎路遥 ｜ 微信：luyao2089 ｜ 官网：luyao2089.cc", columns), "gray"))
        lines.append(colorize(_fit("版权所有 © 黎路遥，保留所有权利", columns), "gray"))
        footer_text = page.footer(self) if page and hasattr(page, "footer") else ""
        footer = footer_text or "↑↓选择 回车执行 ←→/数字键切页 9退出页 Ctrl+C直接退出"
        lines.append(colorize(_fit(footer, columns), "gray"))
        lines.append("─" * columns)
        return lines

    def render(self) -> None:
        if not self.running:
            return
        frame_lines = self.build_frame()
        output = "\x1b[H\x1b[J" + "\r\n".join(frame_lines)
        if output == self._last_output:
            # 内容未变化时跳过写入，避免同一秒内重复清屏打断选择。
            return
        self._last_output = output
        self.output.write(output)
        try:
            self.output.flush()
        except (AttributeError, OSError):
            pass

File: 2.发票自动化/共享CLI/tui/tuiApp.js
// TUI 核心框架：备用屏幕、原始按键模式、帧渲染调度、全局按键分发。
// 页面只负责提供“内容行数组”和“按键处理”，不直接触碰终端底层。
const { ANSI代码, 移动光标到, 清空当前行, 着色 } = require("./ansi");
const { 适配宽度, 右侧补齐 } = require("./width");
const { 格式化时钟 } = require("./format");

function 解析转义键(缓冲内容) {
  // 把常见终端转义序列收敛成统一的按键名，未知序列直接丢弃避免误触。
  if (缓冲内容 === "\x1b") {
    return null;
  }
  if (缓冲内容 === "\x1b[A" || 缓冲内容 === "\x1bOA") {
    return "up";
  }
  if (缓冲内容 === "\x1b[B" || 缓冲内容 === "\x1bOB") {
    return "down";
  }
  if (缓冲内容 === "\x1b[C" || 缓冲内容 === "\x1bOC") {
    return "right";
  }
  if (缓冲内容 === "\x1b[D" || 缓冲内容 === "\x1bOD") {
    return "left";
  }
  if (缓冲内容 === "\x1b[H" || 缓冲内容 === "\x1b[F" || 缓冲内容 === "\x1b[1~" || 缓冲内容 === "\x1b[4~") {
    return 缓冲内容 === "\x1b[1~" || 缓冲内容 === "\x1b[H" ? "home" : "end";
  }
  if (缓冲内容 === "\x1b[5~") {
    return "pgup";
  }
  if (缓冲内容 === "\x1b[6~") {
    return "pgdn";
  }
  if (缓冲内容 === "\x1b[3~") {
    return "delete";
  }
  if (缓冲内容 === "\x1b[Z") {
    return "tab-back";
  }
  if (/^\x1b\[\d+(;\d+)*[A-Za-z~]$/.test(缓冲内容)) {
    return "unknown";
  }
  if (缓冲内容.length > 12) {
    return "unknown";
  }
  return null;
}

function 翻译按键字符(字符) {
  if (字符 === "\r" || 字符 === "\n") {
    return "enter";
  }
  if (字符 === "\x03") {
    return "ctrl-c";
  }
  if (字符 === "\x7f" || 字符 === "\x08") {
    return "backspace";
  }
  if (字符 === "\t") {
    return "tab";
  }
  if (字符 === "\x00") {
    return null;
  }
  return 字符;
}

class TUI应用 {
  constructor(选项 = {}) {
    this.title = 选项.title || "发票自动化";
    this.pages = 选项.pages || [];
    this.currentPageIndex = 0;
    this.statusBarProvider = 选项.statusBarProvider || (() => []);
    this.footerProvider = 选项.footerProvider || (() => "");
    this.onExitRequest = 选项.onExitRequest || (() => {});
    this.onGlobalKey = 选项.onGlobalKey || null;
    this.running = false;
    this.terminalStarted = false;
    this.renderQueued = false;
    this.escapeBuffer = "";
    this.escapeTimer = null;
    // 输出流可注入（测试用 mock），默认写真实终端。
    this.output = 选项.output || process.stdout;
  }

  get page() {
    return this.pages[this.currentPageIndex] || null;
  }

  get columns() {
    return Math.max(48, this.output.columns || process.stdout.columns || 80);
  }

  get rows() {
    return Math.max(14, this.output.rows || process.stdout.rows || 24);
  }

  get contentHeight() {
    // 帧布局固定为：标题 1 + 状态 2 + 菜单 1 + 分隔 1 + 联系方式 1 + 版权 1 + 页脚 1 + 底边 1 = 9 行。
    return Math.max(4, this.rows - 9);
  }

  // 差分渲染的上一帧行缓存；行数变化（终端缩放）时触发整屏重绘。
  lastFrameLines = null;

  切换页面(索引) {
    const 目标 = Number(索引);
    if (Number.isInteger(目标) && 目标 >= 0 && 目标 < this.pages.length) {
      this.currentPageIndex = 目标;
      if (typeof this.page.onEnter === "function") {
        this.page.onEnter(this);
      }
      this.requestRender();
    }
  }

  start() {
    if (this.running) {
      return;
    }
    if (typeof process.stdin.setRawMode === "function") {
      process.stdin.setRawMode(true);
    }
    process.stdin.resume();
    this.running = true;
    this.terminalStarted = true;
    this.lastFrameLines = null;
    // 进入备用屏幕并禁用自动换行：任何一行恰好写满列宽都不会被终端折行，避免画面上下跳动。
    this.output.write(
      ANSI代码.enterAltScreen + ANSI代码.clearScreen + ANSI代码.hideCursor + "\x1b[?7l"
    );
    process.stdin.on("data", (数据块) => this.消费输入(数据块));
    this.resizeHandler = () => this.requestRender();
    this.output.on("resize", this.resizeHandler);
    if (typeof this.page?.onEnter === "function") {
      this.page.onEnter(this);
    }
    this.requestRender();
  }

  stop() {
    if (!this.terminalStarted) {
      return;
    }
    this.running = false;
    if (this.escapeTimer) {
      clearTimeout(this.escapeTimer);
      this.escapeTimer = null;
    }
    this.escapeBuffer = "";
    process.stdin.removeAllListeners("data");
    if (typeof this.output.removeListener === "function" && this.resizeHandler) {
      this.output.removeListener("resize", this.resizeHandler);
      this.resizeHandler = null;
    }
    this.output.write(
      ANSI代码.reset + "\x1b[?7h" + ANSI代码.showCursor + ANSI代码.leaveAltScreen
    );
    if (process.stdin.isTTY) {
      process.stdin.setRawMode(false);
    }
    process.stdin.pause();
    this.terminalStarted = false;
  }

  requestRender() {
    if (this.renderQueued || !this.running) {
      return;
    }
    this.renderQueued = true;
    setImmediate(() => {
      this.renderQueued = false;
      this.render();
    });
  }

  消费输入(数据块) {
    const 文本 = 数据块.toString("utf8");
    for (const 字符 of 文本) {
      if (this.escapeBuffer) {
        if (this.escapeTimer) {
          clearTimeout(this.escapeTimer);
          this.escapeTimer = null;
        }
        this.escapeBuffer += 字符;
        const 解析结果 = 解析转义键(this.escapeBuffer);
        if (解析结果) {
          this.escapeBuffer = "";
          this.分发按键(解析结果);
        } else if (解析结果 === null && this.escapeBuffer.length > 12) {
          this.escapeBuffer = "";
        }
        continue;
      }
      if (字符 === "\x1b") {
        this.escapeBuffer = "\x1b";
        // 单独按 Esc 不会再收到后续字符；给方向键序列留出极短拼接窗口后，将其识别为 esc。
        this.escapeTimer = setTimeout(() => {
          if (this.escapeBuffer === "\x1b") {
            this.escapeBuffer = "";
            this.escapeTimer = null;
            this.分发按键("esc");
          }
        }, 30);
        continue;
      }
      const 按键 = 翻译按键字符(字符);
      if (按键) {
        this.分发按键(按键);
      }
    }
  }

  分发按键(按键) {
    if (!按键 || 按键 === "unknown") {
      return;
    }

    if (按键 === "ctrl-c") {
      this.onExitRequest();
      return true;
    }

    if (this.page && typeof this.page.handleKey === "function") {
      if (this.page.handleKey(按键, this) === true) {
        this.requestRender();
        return true;
      }
    }

    if (typeof this.onGlobalKey === "function" && this.onGlobalKey(按键, this) === true) {
      this.requestRender();
      return true;
    }

    // 左右键在主菜单循环切页：页面内部已消费的按键不会走到这里。
    if (按键 === "left" || 按键 === "right") {
      const 方向 = 按键 === "right" ? 1 : -1;
      this.切换页面((this.currentPageIndex + 方向 + this.pages.length) % this.pages.length);
      return true;
    }

    const 数字按键 = Number(按键);
    if (Number.isInteger(数字按键) && 数字按键 >= 1 && 数字按键 <= this.pages.length) {
      this.切换页面(数字按键 - 1);
      return true;
    }

    if (按键 === "q") {
      this.切换页面(0);
      return true;
    }

    return false;
  }

  构建帧() {
    const 列数 = this.columns;
    const 行数 = this.rows;
    const 内容高度 = this.contentHeight;
    const 时钟 = 格式化时钟();
    const 页面标题 = this.page ? `${this.page.key}.${this.page.title}` : "";

    const 行列表 = [];

    // 标题栏：左侧标题 + 右侧完整时钟（时钟固定 19 字符）。
    const 标题栏 = ` ${this.title}  ${页面标题 ? `│ ${页面标题}` : ""}`;
    行列表.push(
      着色(适配宽度(标题栏, 列数 - 21), "brightCyan") +
        着色(适配宽度(` ${时钟}`, 21), "gray")
    );
    // 状态栏：固定输出两行，保证内容区高度不变。
    const 状态行列表 = this.statusBarProvider(this) || [];
    for (let 索引 = 0; 索引 < 2; 索引 += 1) {
      行列表.push(状态行列表[索引] === undefined ? "" : 状态行列表[索引]);
    }
    // 菜单栏
    行列表.push(this.构建菜单栏(列数));
    // 分隔线
    行列表.push(着色("─".repeat(列数), "gray"));

    // 内容区
    let 内容行列表 = [];
    if (this.page && typeof this.page.render === "function") {
      内容行列表 = this.page.render(this) || [];
    }

    for (let 索引 = 0; 索引 < 内容高度; 索引 += 1) {
      const 内容行 = 内容行列表[索引];
      // 页面内容按电话分析项目的约定不做二次截断：表格列已经在页面层完成统一布局，
      // 这里再次截断会把最后一列切掉，甚至让用户误以为字段没有对齐。
      行列表.push(内容行 === undefined ? 右侧补齐("", 列数) : 右侧补齐(内容行, 列数));
    }

    // 作者与版权信息：放在公共外框中，所有页面统一显示。
    行列表.push(着色(适配宽度("作者：黎路遥 ｜ 微信：luyao2089 ｜ 官网：luyao2089.cc", 列数), "gray"));
    行列表.push(着色(适配宽度("版权所有 © 黎路遥，保留所有权利", 列数), "gray"));

    // 页脚
    let 页脚文本 = "";
    if (this.page && typeof this.page.footer === "function") {
      页脚文本 = this.page.footer(this) || "";
    }
    const 页脚 = 页脚文本 || "↑↓选择 回车执行 ←→切页 数字键切页 q返回总览 Ctrl+C退出";
    行列表.push(着色(适配宽度(页脚, 列数), "gray"));
    行列表.push("─".repeat(列数));

    return 行列表;
  }

  构建菜单栏(列数) {
    const 片段列表 = this.pages.map((页面对象, 索引) => {
      const 是否当前页 = 索引 === this.currentPageIndex;
      const 标签 = `${索引 + 1}${页面对象.title}`;
      const 单元格 = ` ${标签} `;
      return 是否当前页 ? 着色(单元格, "reverse") : 着色(单元格, "brightBlue");
    });
    return 适配宽度(片段列表.join(""), 列数);
  }

  render() {
    if (!this.running) {
      return;
    }
    const 帧行列表 = this.构建帧();
    const 列数 = this.columns;
    const 行列表 = 帧行列表.map((行) => 右侧补齐(行, 列数));

    // 首帧或行数变化（终端缩放）时整屏重绘；否则只重绘变化的行，避免每秒整屏闪烁。
    const 需要整屏重绘 = !this.lastFrameLines || this.lastFrameLines.length !== 行列表.length;

    let 输出 = "";
    if (需要整屏重绘) {
      输出 += ANSI代码.cursorHome + 行列表.join("\r\n");
    } else {
      const 变化列表 = [];
      for (let 索引 = 0; 索引 < 行列表.length; 索引 += 1) {
        if (行列表[索引] !== (this.lastFrameLines[索引] || "")) {
          变化列表.push(移动光标到(索引 + 1, 1) + ANSI代码.reset + 清空当前行() + 行列表[索引]);
        }
      }
      if (变化列表.length > 0) {
        输出 += 变化列表.join("");
      }
    }

    this.lastFrameLines = 行列表;
    if (!输出) {
      return;
    }
    this.output.write(输出);
  }
}

module.exports = {
  TUI应用,
  解析转义键,
  翻译按键字符,
};
